nova/gce/driver.py

```diff
@@ -0,0 +1,807 @@
+
+    def list_instances(self):
+        """
+        Return the names of all the instances known to the virtualization
+        layer, as a list.
+        """
+        compute, project, zone = self.gce_svc, self.gce_project, self.gce_zone
+        all_instances = gceutils.list_instances(compute, project, zone)
+        self._uuid_to_gce_instance.clear()
+        instance_ids = []
+        for instance in all_instances:
+            if instance.state in [
+                    "PROVISIONING", "STAGING", "STOPPING", "SUSPENDING",
+                    "TERMINATED"
+            ]:
+                continue
+            openstack_id = gceutils.get_instances_metadata_key(
+                compute, project, zone, instance, 'openstack_id')
+            if openstack_id is None:
+                openstack_id = self._get_uuid_from_gce_id(instance['name'])
+
+            self._uuid_to_gce_instance[openstack_id] = instance
+
+            instance_ids.append(openstack_id)
+
+        return instance_ids
+
+    def plug_vifs(self, instance, network_info):
+        """Plug VIFs into networks."""
+        pass
+
+    def unplug_vifs(self, instance, network_info):
+        """Unplug VIFs from networks."""
+        pass
+
+    def spawn(self,
+              context,
+              instance,
+              image_meta,
+              injected_files,
+              admin_password,
+              network_info=None,
+              block_device_info=None):
+        """Create a new instance/VM/domain on the virtualization platform.
+        Once this successfully completes, the instance should be
+        running (power_state.RUNNING).
+
+        If this fails, any partial instance should be completely
+        cleaned up, and the virtualization platform should be in the state
+        that it was before this call began.
+
+        :param context: security context <Not Yet Implemented>
+        :param instance: nova.objects.instance.Instance
+                         This function should use the data there to guide
+                         the creation of the new instance.
+        :param image_meta: image object returned by nova.image.glance that
+                           defines the image from which to boot this instance
+        :param injected_files: User files to inject into instance.
+        :param admin_password: set in instance. <Not Yet Implemented>
+        :param network_info:
+           :py:meth:`~nova.network.manager.NetworkManager.get_instance_nw_info`
+        :param block_device_info: Information about block devices to be
+                                  attached to the instance.
+        """
+        """
+        Steps to spwan instances:
+        1. Get GCE image info
+        2. Get appropriate network info
+        3. Instance security groups
+        4. Add ssh keys if required
+
+        """
+        compute, project, zone = self.gce_svc, self.gce_project, self.gce_zone
+        instance_name = instance['name']
```

Comment:
Since Platform9 is upgrading OpenStack to Newton release it is better to . notation instead of [].
Please change the line to instance.name. This should also work with Liberty code.